### PR TITLE
Allow combining Docker task inputs

### PR DESCRIPTION
**Pull request by vreuter**
Not super necessary, but the empty and combine ops seemed natural when I went to make a Docker task input in a notebook. It doesn't feel common, but I could see a user having JSON's of different inputs stored locally and/or fetched from remote stores and then wanting to combine them and this would help w/ that.
- [x] unit/laws for Monoid
- [x] integration
- ~~use Monoid in `TestFunflow`~~
- [x] [re-hide constructors](https://github.com/tweag/funflow2/pull/88#pullrequestreview-664491488)

### CODE REVIEW

**Review by dorranh — CHANGES_REQUESTED** [review]
Thanks for the PR! The Monoid instance seems like it would be especially helpful. I am a bit curious about how the SemiGroup instance would behave in cases where not all of the arguments are specified in the DockerTaskConfig (i.e. you have extra fields after combining the two inputs which are not used by the DockerTask). I think* that it shouldn't cause any errors, but am not 100% sure. We could add a test for this in the integration test suite which now resides under `funflow/test/flows`.

It could also be good to add a small unit test for the logic for `<>`. Would you have time to take a shot at that? If so, I went ahead and added a skeleton where the unit tests can be added [here](https://github.com/tweag/funflow2/pull/88/commits/ff23865b9f6bc8a0baad684bfbf417f3b56d95d1#diff-304e1adfecdf6d2fd24860244727704fa2425d7679c211272b901a1398963096R8).

If building with nix, you can run and build them with:

`nix-build -A funflow-unit-tests && ./result/bin/unit-tests`. The current skeleton should fail to run.

**Comment by vreuter** [comment]
Hey @dorranh I'm testing out the instances, but need to figure out why generating the arbitrary digest seems to be hanging the tests (I added a TODO about that in most recent commit), so this isn't ready for merge yet.

**Comment by vreuter** [comment]
Also, I exposed a couple constructors, mainly for testing though it also didn't seem harmful since the fields of the relevant types already seem to have the domain restriction / safety at the type level (like `CS.Item`), and I couldn't get an alternative approach working (see #91). Also added a couple derived `Show` instances related to usage `QuickCheck` (though also may have benefit in their own right, #86), is that OK?

**Comment by vreuter** [comment]
OK, hanging issue for the associativity test for semigroup seems OK now, taking a different route to an arbitrary `Digest SHA256`

**Review by dorranh — COMMENTED** [review]
Hey! So I would actually say that while inconvenient, we should probably aim to avoid exposing these constructors. The `CAS` library has been around for a while (it's a copy-paste from funflow1) and keeps those constructors private for a reason, the reason being that it is designed around the assumption that it is managing the `Item`s in the content store. An alternative approach could be to do something similar to what we do in the flow tests where we set up a temporary directory for testing: https://github.com/tweag/funflow2/blob/74853d38c77e84eebab851a07858eaa2e63c23f4/funflow/test/TestFunflow.hs#L60. 

I think* that you can set this up once before the property tests start and use `withResource` to pass it into your tests (https://hackage.haskell.org/package/tasty-1.4.1/docs/Test-Tasty.html#g:5). 

It does feel a bit clunky this way, but we are already using this approach in other tests and it allows us to avoid making changes to `CAS`.

**Comment by dorranh** [comment]
PS: In case you haven't seen this already: we also publish the docs for `CAS` on the website: https://tweag.github.io/funflow2/api/cas-store/html/Data-CAS-ContentStore.html#g:1

**Comment by vreuter** [comment]
Cool I will re-hide the constructors and update tests accordingly.

**Comment by dorranh** [comment]
Awesome, thanks! Let me know if you run into any issues

**Review by dorranh — COMMENTED** [review]
Thanks for the updates @vreuter, in general LGTM. Just a couple of small comments / thoughts:

As far as I can tell, what we are doing here is what I would consider a more thorough approach where on each test iteration we generate `DockerTaskInput` to specifically trigger the the different cases for things like de-duplicating the volume mounts. In principle this should work for what we want to test.

 Another approach to this would be to construct a single generator which constructs both kinds of cases (i.e. inputs with and without duplicate mount points) and then just assert 1. that the monoid properties as we do now (in the "pure" assertions) and 2. the property that no duplicate mount points exist in the output. The advantages of this approach would be that I think it should be a bit more efficient since you are only calling the random generator once up-front instead of having to generate the `DockerTaskInputs` in each of the property assertions and that it might require less code.

I'm happy either way, and we can go ahead and merge this as-is since it does test what we want it to. Just something to think about

File: funflow/src/Funflow/Tasks/Docker.hs [hunk 1]
```diff
@@ -61,11 +64,23 @@ data DockerTaskInput = DockerTaskInput
...
-  }
+    argsVals :: Map.Map String Text
+  } deriving (Eq, Show)
+
+instance Semigroup DockerTaskInput where
+  DockerTaskInput{ inputBindings = vols1, argsVals = args1 } <> DockerTaskInput{ inputBindings = vols2, argsVals = args2} = 
+    -- TODO: Better to error for duplicate mounts? This treats merge like normal map.
+    --       An advantage here is that we get "priority list"-like behavior like may be intuitive defining configs.
+    let agg (ms, vs) v = if Set.member (mount v) ms then (ms, vs) else (Set.insert (mount v) ms, v:vs)
+        combVols       = reverse . snd $ foldl' agg (Set.empty, []) (vols1 ++ vols2)
```
[the remark this answers is not in the document]
Comment (vreuter) [reply funflow/src/Funflow/Tasks/Docker.hs:75]: Good call!

**Comment by vreuter** [comment]
> As far as I can tell, what we are doing here is what I would consider a more thorough approach where on each test iteration we generate `DockerTaskInput` to specifically trigger the the different cases for things like de-duplicating the volume mounts. In principle this should work for what we want to test.

Yep, that's what I was aiming for.

> Another approach to this would be to construct a single generator which constructs both kinds of cases (i.e. inputs with and without duplicate mount points) and then just assert 1. that the monoid properties as we do now (in the "pure" assertions) and 2. the property that no duplicate mount points exist in the output. The advantages of this approach would be that I think it should be a bit more efficient since you are only calling the random generator once up-front instead of having to generate the `DockerTaskInputs` in each of the property assertions and that it might require less code.

I agree on at least a couple fronts:
1. It would more efficient (specifically, fewer inputs generated) if we grouped the assertions, like left/right identities and associativity, all under one property, sth like `prop_DockerTaskInputMonoidLawsHold`.
2. It would probably wind up being a bit fewer lines.

I think though that as-is has a couple advantages also:
1. Should a law break, with the properties separated it's easier to quickly see which it is.
2. Separating the case generation out by property ensures each atomic property gets plenty of work, without having a bit trickier code to write a single generator with controlled imbalance to make sure the rarer cases (like certain overlap b/w maps) come up sufficiently often

**Comment by vreuter** [comment]
OK I think this is ready for merge now @dorranh , pending what you think about wording in the docs for the new instances.

File: funflow/src/Funflow/Tasks/Docker.hs [hunk 2]
```diff
@@ -61,11 +64,29 @@ data DockerTaskInput = DockerTaskInput
...
+
+-- | Combine task inputs by combining the collections they contain.
+--
+-- This treats @inputBindings@ as a "priority list"-like structure in case of repeat occurrence 
+-- of the same mount path. Specifically, the item associated with a particular path will be the 
+-- first one, left-to-right, from the @inputBinding@s being combined. Analogous for @argsVals@, 
+-- in accordance with ordinary @Semigroup Map@.
+instance Semigroup DockerTaskInput where
+  DockerTaskInput{ inputBindings = vols1, argsVals = args1 } <> DockerTaskInput{ inputBindings = vols2, argsVals = args2} = 
+    -- TODO: Better to error for duplicate mounts? This treats merge like normal map.
```
Comment (dorranh) [inline funflow/src/Funflow/Tasks/Docker.hs:78]:
```suggestion

```
I think we can remove this TODO since we have the current behavior documented.

**Review by dorranh — APPROVED** [review]
Nice, just updated the CI pipeline to call the new test suite. LGTM

**Pull request merged**